src/server/modules/bulk/services/BulkService.ts

```diff
@@ -0,0 +1,86 @@
+import { injectable } from 'inversify'
+import paparse from 'papaparse'
+import * as interfaces from '../interfaces/BulkService'
+
+import {
+  bulkUploadMaxNum,
+  bulkUploadRandomStrLength,
+  ogHostname,
+} from '../../../config'
+import { BULK_UPLOAD_HEADER } from '../../../../shared/constants'
+import { BulkUrlMapping } from '../../../repositories/types'
+import * as validators from '../../../../shared/util/validation'
+import generateShortUrl from '../../../util/url'
+
+const BULK_UPLOAD_RANDOM_STRING_LENGTH = bulkUploadRandomStrLength
+const BULK_UPLOAD_MAX_NUM = bulkUploadMaxNum
+
+@injectable()
+export class BulkService implements interfaces.BulkService {
+  parseCsv: (dataString: string) => interfaces.CSVSchema = (dataString) => {
+    const schema = {
+      rows: 0,
+      isValid: true,
+      longUrls: [],
+    } as interfaces.CSVSchema
+
+    paparse.parse(dataString, {
+      skipEmptyLines: false,
+      delimiter: ',',
+      step(step, parser) {
+        const rowData = step.data as string[]
+        let validRow = true
+        if (schema.rows === 0) {
+          // if header is invalid
+          if (rowData[0] !== BULK_UPLOAD_HEADER) {
+            schema.isValid = false
+            parser.abort()
+          }
+        } else {
+          const acceptableLinkCount = schema.rows <= BULK_UPLOAD_MAX_NUM + 1 // rows include header
+          const onlyOneColumn = rowData.length === 1
+          const isNotBlacklisted = !validators.isBlacklisted(rowData[0])
+          const isNotEmpty = rowData[0].length > 0
+          const isHttps = validators.isHttps(rowData[0])
+          const validCharacters = validators.isPrintableAscii(rowData[0])
```

Comment:
could we switch to using `isValidUrl` instead of `isPrintableAscii`? it doesn't seem like `isPrintableAscii` should be used for this purpose!

Reply:
@jimvae can comment why we chose this? I'm not sure about the logic for this validation

Reply:
@gweiying @halfwhole Thanks for pointing this out! 

Initially did my own research on url validation and thought that we don't include non ascii characters inside long urls. Just checked that we allow non-ascii characters inside longurls (e.g. mandarin characters etc..) and support their encoding. Also I'll remove validators.isHttps since validators.isValidUrl is the one used in [url.longUrl validation](https://github.com/opengovsg/GoGovSG/blob/feat/bulk-backend/functionality/src/server/models/url.ts#:~:text=if%20(!-,isValidUrl,-(url%2C )

fixes:
- [ ] replace validators.isPrintableAscii and validators.isHttps with validators.[isValidUrl](https://github.com/opengovsg/GoGovSG/blob/feat/bulk-backend/functionality/src/shared/util/validation.ts#:~:text=export%20function-,isValidUrl,-(url%3A )

Reply:
thanks @jimvae